### PR TITLE
CPAL

### PR DESCRIPTION
**Pull request by borisdev**
# Causal Program-aided Language (CPAL)

Twitter handle: @boris_dev

## PAL versus CPAL

In `chains/pals`, PAL stands for Program-aided Langauge. In this contribution as `chains/cpal`, CPAL stands for Causal Program-aided Language. The difference in name is the word "Causal".

Like [PAL](https://arxiv.org/abs/2211.10435), CPAL intends to reduce large language model (LLM) hallucination. 

The CPAL chain is different from the PAL chain because it adds a causal structure (or DAG) to link entity actions (or math expressions). As a side-note, `chains/graph_qa` chain also extracts entity-action-entity relations into a DAG, but it does not connect them to math expressions, not does it sort them in order of dependence.

##

## Evaluation

Preliminary evaluation on simple math word problems shows that this CPAL chain generates less hallucination than the PAL chain because the answers to queries must adhere to the narrative's causal structure.

## Potential "Plan as Code" applications

A Causal Program-aided Language (CPAL) chain allows a user to ask hypothetical questions about a causal narrative. Without the CPAL chain's interface the LLM would still be able to answer such hypothetical questions, but it would hallucinate more frequently than otherwise. Here is an explanation of an upcoming demo app: [Causal Program-aided Language "Plan as Code" demo.](https://github.com/borisdev/cpal-llm-chain-demo).

### Two for one coupon

The thesis here is that the CPAL chain allows you to represent a project plan both as code and as a narrative, giving you the best of both worlds.

### Why Plan as Narrative?

Narratives are the top tool to persuade others of a plan during consensus building. You use a narrative to defend the validity of your plan by illustrating how a causal chain of events can lead to the desired outcome. Narratives are the top medium for ordinary people to convey their causal mental model.

### Why Plan as Code?

Though fast, narratives are problematic as their complexity increases. The problem is LLMs and humans are prone to hallucination when predicting the outcomes of a narrative. The cost of building a consensus around the validity of a narrative outcome grows as its narrative complexity increases. This is a culprit in the "tribal knowledge" problem and "highest paid person in the room" problem. The Amazon-6-pager narrative meeting form attacks this problem. Likewise, the Plan as Code concept attacks this problem. Code does not require tribal knowledge or social power to validate. As narrative complexity increases the value of representing a plan as code goes up. Code is testable, complex narratives are not.

Also, code is quickly composable, complex narratives are not. Composability means it can be integrated with other project plans and applications. For community voting and funding, composable plans can be integrated with a Dapp. For stochastic simulations, a composable plan can be integrated with the DoWhy library.

In summary, a code representation makes a project plan composable and testable.

## Who can review?

Community members can review the PR once tests pass. Tag maintainers/contributors who might be interested:

### CODE REVIEW

**Comment by borisdev** [comment]
moved my CPAL code to my master
